### ISSUE TITLE
Configure multiple providers

### ISSUE THREAD

**Issue by pellared** [open-telemetry/opentelemetry-configuration#5]
It SHOULD be possible to configure multiple trace/meter/logger providers.

Reference https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#tracerprovider

> Notwithstanding any global `TracerProvider`, some applications may want to or
have to use multiple `TracerProvider` instances,
e.g. to have different configuration (like `SpanProcessor`s) for each
(and consequently for the `Tracer`s obtained from them),
or because its easier with dependency injection frameworks.
Thus, implementations of `TracerProvider` SHOULD allow creating an arbitrary
number of `TracerProvider` instances.

**Comment by pellared** [comment]
Just to make it clear. I think that the schema MUST accept multiple providers. Ten depending on the Configurer implementation it will be respected (or not).

**Comment by pellared** [comment]
> @pellared why when you can create multiple providers by running multiple configurers?

I do not know the exact use cases but I already heard that some people are having e.g. 2 trace providers. I have quickly found on thread [here](https://cloud-native.slack.com/archives/C01N3BC2W7Q/p1681913779859649).

**Comment by pellared** [comment]
> This workflow is minimally different than simply having separate config files

Alternatively `OTEL_CONFIG_FILE` env var might need to support multiple file paths.

How will we define that a provider has to be set as a "global provider"?

**Comment by pellared** [comment]
> Multiple tracer providers in one file requires that they be named / identified and that the caller has some way to obtain the instance they want.

1. The names/IDs may not be  needed unless we have some cross-cutting concerns between other elements of the config file (e.g. corelating a trace provider with a tracing instrumentation library). 
2. The "caller" needs to only know which provider should be set a global provider. E.g.

**Comment by jack-berg** [comment]
The spec says it should be possible to create multiple providers, but doesn't give any mechanism for identifying these providers or automatically configuring them. That's new to this proposal.

If multiple providers are possible and can be automatically configured via `OTEL_CONFIG_FILE`, then all instrumentation would have to be aware of this and decide which provider they want:
```
Map<String, TracerProvider> tracerProviders = Configuration.configure(System.getEnv("OTEL_CONFIG_FILE")) .. // Init multiple tracer providers from OTEL_CONFIG_FILE
  .getTracerProviders();
HttpServerInstrumentation.create(tracerProviders.get("tracer-provider1")); // Initialize http instrumentation with "tracer-provider-1"
DbInstrumentation.create(tracerproviders.get("tracer-provider2"))); // Initializer db instrumentation with "tracer-provider-2"
```

With one provider per file, its still possible to have multiple providers. The caller is just responsible for referencing each configuration file and passing the resulting providers to the appropriate place in the application:
```
TracerProvider provider1 = Configuration.configure("/config1.yaml").getTracerProvider();
TracerProvider provider2 = Configuration.configure("/config2.yaml").getTracerProvider();

HttpServerInstrumentation.create(provider1);
DbInstrumentation.create(provider2);
```

> The names/IDs may not be needed unless we have some cross-cutting concerns between other elements of the config file (e.g. corelating a trace provider with a tracing instrumentation library).

If one tracer provider is the global, and that's indicated in the config file, and instrumentations don't select which provider they want because presumably they choose the global, then how are the non-global providers used?

**Comment by pellared** [comment]
> The spec says it should be possible to create multiple providers, but doesn't give any mechanism for identifying these providers or automatically configuring them. That's new to this proposal.

Correct. I think this is something that would part of "Configurer/Configuration API/structure". It does not have to be part of Traces/Metrics/Logs API. This is only something that needs to be parsed/processed during "telemetry pipeline" setup.

**Comment by pellared** [comment]
Personally, I do not want to propose not decide "how" to do it. My proposals were just "drafts" to "visualize" the issue.

First of all, we should decide if this is something that is planned be addressed.

In my opinion, the Configuration Model MUST allow instantiating multiple providers of the same type to allow complex configurations. One of the reasons we want to use the Configuration Model is to allow setting up complex things which would be almost impossible using env vars.

**Comment by jack-berg** [comment]
> In my opinion, the Configuration Model MUST allow instantiating multiple providers of the same type to allow complex configurations. 

That requirement is satisfied by the ability to have multiple configuration files / models. It's a great simplifying assumption to say that the configuration model defines one tracer provider / meter provider / logger provider configuration.

A user that insists on putting multiple configurations in a single file can always use the YAML syntax to define multiple documents in a single file:
```
---
resource: ...
tracer_provider: ...
---
resource: ...
tracer_provider: ...
```
And parse like:
```
List<TracerProvider> tracerProviders = ParseDocuments(new File("/multi-config.yaml"))
    .stream()
    .map(document -> Configuration.configure(document).getTracerProvider())
    .collect(toList());

TracerProvider provider1 = tracerProviders.get(0);
TracerProvider provider2 = tracerProviders.get(1);
```

Having multiple providers is an exceptional case. The [link](https://cloud-native.slack.com/archives/C01N3BC2W7Q/p1681913779859649) you posted reiterates that. We don't need to burden application owners with this detail, the vast majority of which will only be confused by why they need to define an array of providers. And we don't need to burden SDK / instrumentation authors with trying to figure out what to when multiple providers are present.

**Comment by MikeGoldsmith** [comment]
I agree with @jack-berg - I think multiple providers per Configuration Model unnecessarily complicates the schema and doesn't help the use case of accessing multiple providers because they cannot easily be accessed.

The example above can return multiple Configuration Models from the same file using multiple YAML documents will work and gives the same index based access returning multiple providers from the same model would.

**Comment by pellared** [comment]
> The example above can return multiple Configuration Models from the same file using multiple YAML documents will work and gives the same index based access returning multiple providers from the same model would.

People using automatic instrumentation would not able to (easily?) profit from such approach.

**Comment by tsloughter** [comment]
Linking with instrumentation libraries?

I should say that having named providers would make life easier in Erlang because I guess we do something similar to automatic instrumentation (even though it doesn't actually instrument anything automatically, it just sets up the providers at boot time so tracers are available before dependencies boot). So the user is usually not going to run anything like `start_tracer_provider` but let it be done on boot based on configuration.

I like the multi-file approach though and was just resigning to let it be that the provided config file at boot will start the global providers and any additional providers would be created by the user by manually calling the configurer and providing names for them (providers are named processes in Erlang) at that time.
